project_cost_report/report.py

```diff
@@ -0,0 +1,436 @@
+        """
+        project = self.get_project_from_report_context(report_context)
+        rendering_variables = self._get_rendering_variables(project, report_context)
+        base_url = self.env['ir.config_parameter'].sudo().get_param('web.base.url')
+        rendering_variables.update({
+            'mode': 'print',
+            'base_url': base_url,
+        })
+        body = self.env['ir.ui.view'].render_template(
+            "project_cost_report.cost_report_pdf",
+            values=rendering_variables,
+        )
+        header = self.env['ir.actions.report'].render_template(
+            "web.internal_layout", values=rendering_variables
+        )
+        header = self.env['ir.actions.report'].render_template(
+            "web.minimal_layout", values=dict(rendering_variables, subst=True, body=header)
+        )
+        return self.env['ir.actions.report']._run_wkhtmltopdf(
+            [body],
+            header=header,
+            landscape=True,
+            specific_paperformat_args={
+                'data-report-margin-top': 10,
+                'data-report-header-spacing': 10
+            }
+        )
+
+    @api.model
+    def get_foldable_categories(self, project_id):
+        """Get a dictionary containing all the foldable categories for the report.
+
+        Here is an example of expected result:
+
+        {
+            'product': [1, 13, 53, 27],
+            'time': [1, 4, 5],
+            'outsourcing': [954, 1058],
+        }
+
+        The IDS contained in each category depends on the model used to group
+        analytic lines in each category.
+
+        :param int project_id: the ID of the project
+        :rtype: Mapping[str, List[int]]
+        """
+        return {}
+
+
+def _group_analytic_lines(lines: models.Model, key: Callable) -> Mapping[object, models.Model]:
+    grouped_lines = {}
+    for line in lines:
+        key_ = key(line)
+        if key_ not in grouped_lines:
+            grouped_lines[key_] = lines.env['account.analytic.line']
+        grouped_lines[key_] |= line
+    return grouped_lines
+
+
+class CostReportCategory:
+    """Cost report category.
+
+    This class is used to simplify the qWeb templates.
+    It contains the data related to a category of the report.
+    """
+
+    def __init__(self, id_: int, name: str, lines: models.Model, folded: bool):
+        """Initialize the report category.
+
+        :param id: the ID of the related product.category record
+        :param name: the name of the category
+        :param lines: a recordset of Odoo account.analytic.line
+        :param folded: whether the category is folded or not
+        """
+        self.id = id_
+        self.name = name
+        self.lines = lines
+        self.folded = folded
+        self.total = float_round(sum(l.amount for l in lines), 2)
+
+
+class ProjectCostReportWithProducts(models.TransientModel):
+
+    _inherit = 'project.cost.report'
+
+    def _get_product_analytic_line_domain(self, project):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=3675949)